Implement torch.igamma

mfkasim1 opened the pull request:
Fixes #41637
This is regularized lower incomplete gamma function, equivalent to scipy's `gammainc` and tensorflow `igamma`.

cc @fritzo @mruberry

mfkasim1 committed:
Add igamma (CPU only for now)

mfkasim1 committed:
Add docs for torch.igamma

mfkasim1 committed:
Add the CUDA implementation

mfkasim1 committed:
Fix bug in cuda undeclared vars

mfkasim1 committed:
Use C10_HOST_DEVICE

mfkasim1 committed:
Add CUDA kernel and edge cases test

mfkasim1 committed:
Derivatives w.r.t. x of igamma

mfkasim1 committed:
Refactor igamma in CUDA

mfkasim1 committed:
Cast to accscalar_t for ::isinf cuda

fritzo commented on test/test_autograd.py:
```diff
@@ -2843,6 +2843,14 @@ def test_pow_scalar_base(self):
+    def test_igamma(self):
+        # 1e-3 offset to avoid zeros
+        # NOTE: derivative for s is not implemented
+        s = (torch.rand(3, dtype=torch.double) + 1e-3)
+        x = (torch.rand(3, dtype=torch.double) + 1e-3).requires_grad_()
```
Consider testing on more points, say 100 points instead of 3. As I understand this is basically free due to tensor overhead.

mfkasim1 replied:
Done. I've added the tests for 100 points and it failed, so I just changed it to follow scipy's implementation (which makes the code substantially larger)

fritzo reviewed (commented):
Tests look good, I would only test on more points.

mfkasim1 committed:
Implement igam like scipy

mfkasim1 committed:
Expand igam to follow scipy and add more tests

mfkasim1 committed:
Fix bug of int to float conv and remove some vars

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+           r"""
+igamma(input, other, *, out=None) -> Tensor
+
+Regularized lower incomplete gamma function
```
"Computes the regularized lower incomplete gamma function:"

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+Regularized lower incomplete gamma function
+
+.. math::
+    \text{out}_{i} = \frac{1}{\Gamma(\text{input})} \int_0^\text{other} t^{\text{input}-1} e^{-t} dt
```
The input and output values in the expression also need subscripts.

mruberry replied:
After this mathematical expression is a good opportunity to describe the constraints on the input, and what will happen if those constraints are violated.

mruberry replied:
This needs to define the Gamma function. It's interesting that SciPy's documentation doesn't seem to define it, either.

mfkasim1 replied:
Is there any docs example in pytorch to explain the constraints?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+.. math::
+    \text{out}_{i} = \frac{1}{\Gamma(\text{input})} \int_0^\text{other} t^{\text{input}-1} e^{-t} dt
+
+The shapes of ``input`` and ``other`` must be
```
This is good but we're starting to use a new style for statements like this. See the "Supports..." statement here for an example: https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+The shapes of ``input`` and ``other`` must be
+:ref:`broadcastable <broadcasting-semantics>`.
+
+.. note::
```
@albanD, what's your recommendation for a statement like this?

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+:ref:`broadcastable <broadcasting-semantics>`.
+
+.. note::
+    The backward pass for :math:`\text{input}` has not been implemented.
```
Instead of a mathematical expression use :attr:\`input\`

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,36 @@ def merge_dicts(*dicts):
+
+""" + r"""
+Args:
+    input (Tensor): the first input tensor
```
This is a good opportunity to reiterate the constraints.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6040,6 +6040,19 @@
 - func: hypot_(Tensor(a!) self, Tensor other) -> Tensor(a!)
   variants: method
 
+- func: igamma.out(Tensor self, Tensor other, *, Tensor(a!) out) -> Tensor(a!)
```
I think you can get rid of these explicit dispatch: entries.

mfkasim1 replied:
Do you mean to remove the lines

    dispatch:
      CPU, CUDA: igamma*

for all igamma functions?

mruberry replied:
Yep.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
       zeta(double(n + 1), x);
 }
 
+// regularized lower incomplete gamma
```
If this code is copied or derived from SciPy we should be sure to cite it properly. See, for example. how we cite Cephes in this same file. 

@rgommers I couldn't find the SciPy license in PyTorch, seems like we should add it here?

rgommers replied:
That's one option, but probably not the optimal one.  I had a look at how all the license info is structured, and there's a few issues with it. The `NOTICE` file contains license info of some vendored code. I think the SciPy license should be added there - that will future-proof things so SciPy code can simply be copied in the future, without having to worry about adding more license info each time.

`NOTICE` should also be referenced from `LICENSE`.

There may be more to do, for example `NOTICE` says that Caffe2 is Apache 2 licensed, which is important and should be reflected in the overall PyTorch license that is advertised in for example the PyPI classifier (which is no longer just BSD, it's hybrid and not GPLv2-compatible - which is luckily no longer very important).

mruberry replied:
Looks like `NOTICE` just contains a few Caffe2 licenses? Looks like @ezyang worked on the file (like nearly every file in PyTorch). @ezyang, is that the right place to stuff the SciPy license?

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
 }
 
+// regularized lower incomplete gamma
+// the regularized lower and upper incomplete gamma follows SciPy's
```
Wait is there going to be an implementation of the regularized upper incomplete gamma function, too? I thought this PR was just the regularized lower incomplete gamma function.

mfkasim1 replied:
The upper incomplete gamma function is required to calculate the lower incomplete gamma function, so I just add it here, but I don't add the upper one as an API.

mruberry replied:
That makes mathematical sense, of course.

As an aside, I don't think I like either the name "igamma" or "gammainc." Both suggest this is the "incomplete Gamma function," which isn't even a thing, and definitely not the regularized lower/upper incomplete Gamma function. SciPy's and Tensorflow's names for the regularized upper incomplete Gamma function seem even sillier. Oh well.

My question then, should this PR also add igammac? Or would you be interested in adding it in a follow-up PR?

mfkasim1 replied:
I'd prefer to finish this first before jumping into `igammac`. It's giving me quite plenty things to learn already.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+// the regularized lower and upper incomplete gamma follows SciPy's
+// implementation
+
+template <typename scalar_t>
```
Add a comment explaining what this function does.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static scalar_t lanczos_sum_expg_scaled(scalar_t x) {
```
SciPy mentions its implementation comes from Boost. We should also populate the correct reference for these values. See https://live.boost.org/doc/libs/1_54_0/boost/math/special_functions/lanczos.hpp.

mfkasim1 replied:
I've added the boost license as well here

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static scalar_t _igam_helper_fac(scalar_t a, scalar_t x) {
```
This function needs a comment, too.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static scalar_t _igam_helper_series(scalar_t a, scalar_t x) {
```
Comment here as well.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static scalar_t _igamc_helper_continued_fraction(scalar_t a, scalar_t x) {
```
Comment here.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static inline scalar_t calc_igammac(scalar_t a, scalar_t x) {
```
Comment here.

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,455 @@ static inline float calc_polygamma(int64_t n, float x) {
+}
+
+template <typename scalar_t>
+static inline scalar_t calc_igamma(scalar_t a, scalar_t x) {
```
Comment explaining the approach to compute this function here.

mruberry commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -394,6 +394,15 @@ struct Vec256 {
     return map(calc_i0);
   }
+  Vec256<T> igamma(const Vec256<T> &x) const {
+    // TODO: is there a way to implement SIMD better than evaluating it
```
I don't think so, no.

mruberry commented on test/test_autograd.py:
```diff
@@ -2843,6 +2843,14 @@ def test_pow_scalar_base(self):
+    def test_igamma(self):
+        # 1e-3 offset to avoid zeros
+        # NOTE: derivative for s is not implemented
+        s = (torch.rand(100, dtype=torch.double) + 1e-3)
```
So `s` doesn't need to be strictly positive, and `x` doesn't need to be weakly positive?

mruberry commented on test/test_torch.py:
```diff
@@ -17489,6 +17493,52 @@ def test_hypot(self, device, dtype):
                 expected = np.hypot(input[0].cpu().numpy(), input[1].cpu().numpy())
             self.assertEqual(actual, expected)
 
+    @dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
```
You also support half on CUDA, so I'd expected `@dtypesIfCUDA(torch.half, ...)` as well.

mruberry commented on test/test_torch.py:
```diff
@@ -17489,6 +17493,52 @@ def test_hypot(self, device, dtype):
+            (torch.rand(10, device=device).to(dtype), torch.rand(10, device=device).to(dtype)),
+            (torch.rand((3, 3, 3), device=device).to(dtype), torch.rand((3, 3, 3), device=device).to(dtype)),
+            (torch.rand((100, 1), device=device).to(dtype), torch.rand((100, 1), device=device).to(dtype).transpose(0, 1)),
+            (torch.randint(100, (10, ), device=device, dtype=torch.long), torch.rand(10, device=device).to(dtype))
```
How does this case work? I thought this function only supported float types?

mruberry replied:
I like that you're programmatically enumerating the inputs and comparing with SciPy, but most of the tested values are in the range [0, 1). Are there other interesting ranges to probe for numerical accuracy?

mfkasim1 replied:
I think it converts the integer to float types. Should I just remove this?

mruberry commented on test/test_torch.py:
```diff
@@ -17489,6 +17493,52 @@ def test_hypot(self, device, dtype):
+            expected = torch.from_numpy(expected).to(dtype)
+            self.assertEqual(actual, expected)
+
+    @dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
```
Same dtype note here as above.

mruberry commented on test/test_torch.py:
```diff
@@ -17489,6 +17493,52 @@ def test_hypot(self, device, dtype):
+
+    @dtypesIfCPU(torch.bfloat16, torch.float32, torch.float64)
+    @dtypes(torch.float32, torch.float64)
+    def test_igamma_edge_cases(self, device, dtype):
```
This is a really cool test.

mfkasim1 replied:
Thanks!

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -174,6 +174,458 @@ static inline __host__ __device__ scalar_t calc_polygamma(int n, scalar_t x) {
   return ((n % 2) ? 1.0 : -1.0) * ::exp(::lgamma(static_cast<scalar_t>(n) + 1.0)) * zeta(static_cast<scalar_t>(n + 1), x);
 }
 
+// regularized lower & upper incomplete gamma
```
Just like the CPU version of this file, ensure that every function you're adding has a comment explaining what it is and, ideally, how to use it.

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -174,6 +174,458 @@ static inline __host__ __device__ scalar_t calc_polygamma(int n, scalar_t x) {
 }
 
+// regularized lower & upper incomplete gamma
+template <typename scalar_t>
```
Are these functions completely duplicative with the CPU header updates?

mfkasim1 replied:
Yes ...

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -174,6 +174,458 @@ static inline __host__ __device__ scalar_t calc_polygamma(int n, scalar_t x) {
+
+template <typename scalar_t>
+static inline __host__ __device__ scalar_t calc_igamma(scalar_t a, scalar_t x) {
+  using accscalar_t = at::acc_type<scalar_t, /*is_cuda=*/true>;
```
This is interesting. Why the decision to use the accumulate type here?

mfkasim1 replied:
Mostly because I just follow the implementation of `digamma`, but when I checked it, it agrees with I want because it converts the `bfloat16` and `half` to `float`, so I don't have to do template specialization for `bfloat16` and `half`.

mruberry replied:
That makes sense. I wonder how numerically stable computing in bfloat16 or half would be, but I think that's an academic question we don't need to try an answer in this PR.

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -174,6 +174,458 @@ static inline __host__ __device__ scalar_t calc_polygamma(int n, scalar_t x) {
+  static accscalar_t LARGERATIO = 4.5;
+
+  // boundary values following SciPy
+  if ((x < 0) || (a < 0)) {
```
Does a have to be strictly positive, so the check would be `(a <= 0)`?

mruberry replied:
I see you have a special case for `a == 0` below, but would appreciate your thinking on it. Or maybe this function needs a comment explaining the special case?

mfkasim1 replied:
I follow the boundary condition from SciPy where it defines `res = 1.0` if `a = 0` and `x > 0`.
I'm not sure if it's mathematically correct, but if we want to define `a` to be strictly positive, I can add another boundary condition to make it `nan` whenever `a == 0`.

mruberry replied:
Following SciPy sounds like a good start. Would you just add a comment to the code explaining that choice?

mfkasim1 replied:
Sure!

mruberry commented:
Hey @mfkasim1, thanks for submitting this PR! It looks really good. I asked a few questions and made some comments/suggestions about the tests and docs. 

Are there parts of the implementation which are especially novel, or is the implementation mostly copied from Boost/SciPy?

mfkasim1 commented:
Hi @mruberry, thanks for your comments and suggestions. I will try to address them in the next commit.
The code is mostly copied from SciPy.

mfkasim1 committed:
Add more comments in the code and add test cases

mfkasim1 committed:
Comply with flake8

mfkasim1 committed:
Fix bug in igamma test

mfkasim1 committed:
Add half support for CPU, but removed half & bhalf for GPU

mfkasim1 commented:
@mruberry Here's the summary of what I changed since you last reviewed:

- put license, comments, and update docs as you suggested
- added `igamma` for half for CPU
- dropped `igamma` for half and bhalf datatype for CUDA, because it seems to always fail the test (although the CPU version passes the test, don't know why)

mfkasim1 commented on tools/autograd/derivatives.yaml:
```diff
@@ -540,6 +540,10 @@
   self: not_implemented("i0")
 
+- name: igamma(Tensor self, Tensor other) -> Tensor
+  self: 'not_implemented("igamma: a")'
```
`a` is the first argument of incomplete gamma, usually written as `igamma(a, x)`.
If you think it causes more confusion, I can change it to `self`.

albanD replied:
The doc calls it "input" and the signature "self". Maybe renaming them all to the same thing will make this clearer indeed.

mruberry replied:
Our idiosyncratic convention is that signature is self and docs are input. It should be referred to using the name in the docs.

mfkasim1 replied:
Removing the single quotes makes the build failed. I guess because there is a space in the string. But I've changed it to `igamma: self` in the latest commit

mruberry replied:
I think it's "igamma: input" here to be consistent with the docs?

mfkasim1 replied:
Ah, yes, sorry, a bit confused. Fixed this.

mfkasim1 committed:
Removed test extreme

mfkasim1 committed:
Update docs

mfkasim1 committed:
Update derivatives error message

mruberry commented:
> * put license, comments, and update docs as you suggested

Let's give @rgommers and @ezyang a chance to make their recommendations here, too.

> * dropped `igamma` for half and bhalf datatype for CUDA, because it seems to always fail the test (although the CPU version passes the test, don't know why)

That's really interesting. Is the failure a precision issue? Is it because the half inputs aren't representable in float? That is, when testing half vs float32 it's important to ensure the values are actually representable in half, too. This can be done by creating the values in half and then converting them to float32.

I'd like to understand what's going on here better and support half on CUDA, too.

mfkasim1 committed:
Update docs and modif to comply with XLA (?)

mfkasim1 committed:
Add dispatch back to native_functions.yaml

mfkasim1 committed:
Merge remote-tracking branch 'upstream/master'

mfkasim1 commented:
> That's really interesting. Is the failure a precision issue? Is it because the half inputs aren't representable in float? That is, when testing half vs float32 it's important to ensure the values are actually representable in half, too. This can be done by creating the values in half and then converting them to float32.

Representable issue should not be a problem here because it passes the test for half in CPU, but failed in CUDA.
It's probably different type of calculations performed in CPU and CUDA. It might be from my implementation (i.e., my CPU and CUDA implementations are slightly different), or from the different implementations of std library and CUDA library.

rgommers commented:
> > * put license, comments, and update docs as you suggested
> 
> Let's give @rgommers and @ezyang a chance to make their recommendations here, too.

I would not recommend inlining full license copies in the middle of a header file (or any code file really), that seems unusual and hard to discover. My first preference would be to add them to a central place (i.e., `NOTICE` in the repo root), my second preference would be at the top of the file in which the code is used.

ezyang commented:
Either of Ralf's suggestions seem fine. Historically, we're a bit bad about putting everything we should in NOTICE, but that's what it's there for.

mfkasim1 commented:
> I would not recommend inlining full license copies in the middle of a header file (or any code file really), that seems unusual and hard to discover. My first preference would be to add them to a central place (i.e., NOTICE in the repo root), my second preference would be at the top of the file in which the code is used.

Thanks! I will put it in the NOTICE.
One more thing, my test always fails in XLA with error message:

    Oct 15 12:28:44 ======================================================================
    Oct 15 12:28:44 ERROR [0.013s]: test_igamma_edge_cases_xla_float64 (__main__.TestTorchDeviceTypeXLA)
    Oct 15 12:28:44 ----------------------------------------------------------------------
    Oct 15 12:28:44 Traceback (most recent call last):
    Oct 15 12:28:44   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
    Oct 15 12:28:44     result = test_fn(self, *args)
    Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 17442, in test_igamma_edge_cases
    Oct 15 12:28:44     calc = torch.igamma(input0, input1)
    Oct 15 12:28:44 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
    Oct 15 12:28:44 
    Oct 15 12:28:44 ======================================================================
    Oct 15 12:28:44 ERROR [6.137s]: test_memory_format_operators_xla (__main__.TestTorchDeviceTypeXLA)
    Oct 15 12:28:44 ----------------------------------------------------------------------
    Oct 15 12:28:44 Traceback (most recent call last):
    Oct 15 12:28:44   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
    Oct 15 12:28:44     result = test_fn(self, *args)
    Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15160, in test_memory_format_operators
    Oct 15 12:28:44     torch.channels_last)
    Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15129, in _test_helper
    Oct 15 12:28:44     result_c = fn(x_c, y_c)
    Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15075, in <lambda>
    Oct 15 12:28:44     lambda x, y: x.igamma(y),
    Oct 15 12:28:44 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage

As I don't have XLA device and don't know why it fails, could anyone please help me with this issue?
Thanks!

JackCaoG commented:
> > I would not recommend inlining full license copies in the middle of a header file (or any code file really), that seems unusual and hard to discover. My first preference would be to add them to a central place (i.e., NOTICE in the repo root), my second preference would be at the top of the file in which the code is used.
> 
> Thanks! I will put it in the NOTICE.
> One more thing, my test always fails in XLA with error message:
> 
> ```
> Oct 15 12:28:44 ======================================================================
> Oct 15 12:28:44 ERROR [0.013s]: test_igamma_edge_cases_xla_float64 (__main__.TestTorchDeviceTypeXLA)
> Oct 15 12:28:44 ----------------------------------------------------------------------
> Oct 15 12:28:44 Traceback (most recent call last):
> Oct 15 12:28:44   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
> Oct 15 12:28:44     result = test_fn(self, *args)
> Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 17442, in test_igamma_edge_cases
> Oct 15 12:28:44     calc = torch.igamma(input0, input1)
> Oct 15 12:28:44 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
> Oct 15 12:28:44 
> Oct 15 12:28:44 ======================================================================
> Oct 15 12:28:44 ERROR [6.137s]: test_memory_format_operators_xla (__main__.TestTorchDeviceTypeXLA)
> Oct 15 12:28:44 ----------------------------------------------------------------------
> Oct 15 12:28:44 Traceback (most recent call last):
> Oct 15 12:28:44   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
> Oct 15 12:28:44     result = test_fn(self, *args)
> Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15160, in test_memory_format_operators
> Oct 15 12:28:44     torch.channels_last)
> Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15129, in _test_helper
> Oct 15 12:28:44     result_c = fn(x_c, y_c)
> Oct 15 12:28:44   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15075, in <lambda>
> Oct 15 12:28:44     lambda x, y: x.igamma(y),
> Oct 15 12:28:44 RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
> ```
> 
> As I don't have XLA device and don't know why it fails, could anyone please help me with this issue?
> Thanks!

@mfkasim1 XLA device does not have storage, you can run your test only on CPU and GPU by adding `@ onlyOnCPUAndCUDA` decorator.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,41 @@ def merge_dicts(*dicts):
+    \text{out}_{i} = \frac{1}{\Gamma(\text{input}_i)} \int_0^{\text{other}_i} t^{\text{input}_i-1} e^{-t} dt
+
+where (1) :attr:`input` and :attr:`other` must be non-negative and (2)
+:attr:`input` and :attr:`other` must not be zero at the same index.
```
The elements may not be at the same index due to broadcasting. I also think you can flip this sentence:

"when both :math:\`\\text{input}_i\` and :math:\`\\text{other}_i\` are weakly positive and at least one is strictly positive. If both are zero or either is negative then :math:\`\\text{out}_i=\\text{nan}`.

mfkasim1 replied:
I'm not sure how to express the index to represent the broadcasting. I looked at `torch.hypot` documentation, it also puts the same indices for both operands. I think users who know broadcast should understand what it really means.

mruberry replied:
I agree that _i is clear because it suggests the correspondence. I meant to say that I think saying the elements have the same index, like the current expression does, is misleading. I think you're right that most readers familiar with PyTorch would know what the current documentation means, but we should strive to be as clear as possible. A lot of our code and documentation is copy-pasted, after all, so imperfections have a habit of multiplying.

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,41 @@ def merge_dicts(*dicts):
+
+.. note::
+    The backward pass with respect to :attr:`input` is not supported.
+    Please open a feature request on github if you need it.
```
"Please open a feature request on github if you need it." -> "Please open an issue on the PyTorch Github to request it."

mfkasim1 replied:
Done in the recent commit

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,41 @@ def merge_dicts(*dicts):
+and float inputs.
+
+.. note::
+    The backward pass with respect to :attr:`input` is not supported.
```
"not supported" -> "not yet supported."

mruberry commented on test/test_torch.py:
```diff
@@ -17373,6 +17377,74 @@ def test_hypot(self, device, dtype):
+                input1 = input1.to(torch.float)
+            expected = scipy.special.gammainc(input0.cpu().numpy(), input1.cpu().numpy())
+            expected = torch.from_numpy(expected).to(dtype)
+            if half_prec:
```
Is this check and its conversions necessary? I thought expected would be converted to the dtype of actual in the previous line.

mfkasim1 replied:
Ah, yes. It's probably from the older version of the commit that I forgot to change. Thanks!

mfkasim1 committed:
Licenses at NOTICE and no XLA tests

mfkasim1 commented:
> @mfkasim1 XLA device does not have storage, you can run your test only on CPU and GPU by adding @ onlyOnCPUAndCUDA decorator.

@JackCaoG thanks for your suggestions. The strange thing is another test: `test_igamma_common` doesn't seem to throw the similar error. But for now, I will just run the test on CPU and CUDA

mfkasim1 commented:
Is there any way to know the environment in [`pr/pytorch-linux-bionic-rocm3.8-py3.6`?](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.8-py3.6-test2/1499/console)
It fails on CUDA float64 test with some unreasonable error, but I can't reproduce it in my CUDA, however I changed the random seed.
Here's the error:

    10:45:41 ======================================================================
    10:45:41 FAIL: test_igamma_common_cuda_float64 (__main__.TestTorchDeviceTypeCUDA)
    10:45:41 ----------------------------------------------------------------------
    10:45:41 Traceback (most recent call last):
    10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 828, in wrapper
    10:45:41     method(*args, **kwargs)
    10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
    10:45:41     result = test_fn(self, *args)
    10:45:41   File "test_torch.py", line 17433, in test_igamma_common
    10:45:41     self._helper_test_igamma(loglo, loghi, device, dtype)
    10:45:41   File "test_torch.py", line 17424, in _helper_test_igamma
    10:45:41     self.assertEqual(actual, expected)
    10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1139, in assertEqual
    10:45:41     self.assertTrue(result, msg=msg)
    10:45:41 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1e-07 and atol=1e-07, found 1564 element(s) (out of 10000) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 2756.12890625 (2757.12890625 vs. 1.0), which occurred at index (51, 8).

mfkasim1 committed:
Removed the XLA tests

mfkasim1 committed:
Use logspace in test to be more deterministic

mfkasim1 committed:
No logspace in Half bug fix

mfkasim1 committed:
Skip ROCm CUDA test

mruberry commented:
> Is there any way to know the environment in [`pr/pytorch-linux-bionic-rocm3.8-py3.6`?](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm3.8-py3.6-test2/1499/console)
> It fails on CUDA float64 test with some unreasonable error, but I can't reproduce it in my CUDA, however I changed the random seed.
> Here's the error:
> 
> ```
> 10:45:41 ======================================================================
> 10:45:41 FAIL: test_igamma_common_cuda_float64 (__main__.TestTorchDeviceTypeCUDA)
> 10:45:41 ----------------------------------------------------------------------
> 10:45:41 Traceback (most recent call last):
> 10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 828, in wrapper
> 10:45:41     method(*args, **kwargs)
> 10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 273, in instantiated_test
> 10:45:41     result = test_fn(self, *args)
> 10:45:41   File "test_torch.py", line 17433, in test_igamma_common
> 10:45:41     self._helper_test_igamma(loglo, loghi, device, dtype)
> 10:45:41   File "test_torch.py", line 17424, in _helper_test_igamma
> 10:45:41     self.assertEqual(actual, expected)
> 10:45:41   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1139, in assertEqual
> 10:45:41     self.assertTrue(result, msg=msg)
> 10:45:41 AssertionError: False is not true : Tensors failed to compare as equal! With rtol=1e-07 and atol=1e-07, found 1564 element(s) (out of 10000) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 2756.12890625 (2757.12890625 vs. 1.0), which occurred at index (51, 8).
> ```

This is a MASSIVE different on ROCm. Skipping the test on ROCm is OK for now. Would you also file an issue saying that this test will fail on ROCm when it goes in, include a description of the test failure, and then cite the issue you file on the same line that skips ROCm:

```
@skipCUDAIfRocm  # See issue (link to issue goes here)
```

mruberry commented on NOTICE:
```diff
@@ -281,6 +281,113 @@ Apache License Version 2.0:
+=======================================================================
+
+   Code derived from implementations in the Cephes Math Library should
+   mention its derivation and reference this note (ex. 'This function is
```
This paragraph needs to be updated if the Cephes license is moved here, because this is no longer a note.

"and reference this note" -> "and reference the following Cephes 3-Clause BSD license. For example, 'This function is derived from the implementation of X in the Cephes Math Library. See the Cephes 3-Clause BSD license in PyTorch's NOTICE file.' The license is:

mruberry commented on NOTICE:
```diff
@@ -281,6 +281,113 @@ Apache License Version 2.0:
+SciPy's 3-Clause BSD License
+=======================================================================
+
+This repo contains SciPy code, which is licensed under 3-Clause BSD
```
You can remove this sentence and just put the license here.

mruberry commented on NOTICE:
```diff
@@ -281,6 +281,113 @@ Apache License Version 2.0:
+Boost's 1.0 Software License
+=======================================================================
+
+This repo contains Boost 1.0 code, with the following license:
```
Analogous note update here

mruberry commented on test/test_torch.py:
```diff
@@ -17373,6 +17375,67 @@ def test_hypot(self, device, dtype):
+        inputs = [
+            (vec1, vec1.transpose(0, 1)),
+            (vec1, vec1),  # for large number, it should approach 0.5
+            (vec1, 0.5 * vec1),  # test for considerable ratio
```
Make the first input discontiguous for this test, too.

mfkasim1 replied:
What do you mean by "discontiguous"?

mruberry replied:
Such that not all the elements of the tensor are sequential in memory. This could be done, for example, by taking every other element of vec1:

```
vec1[::2]
```

mfkasim1 replied:
Ah, sure!

mruberry commented on test/test_torch.py:
```diff
@@ -17373,6 +17375,67 @@ def test_hypot(self, device, dtype):
+            (vec1, vec1.transpose(0, 1)),
+            (vec1, vec1),  # for large number, it should approach 0.5
+            (vec1, 0.5 * vec1),  # test for considerable ratio
+            (vec1, 2.0 * vec1),
```
Add two test cases where each input is a scalar:

```
(.378, vec1),
(vec, .018)
```

Or similar.

mfkasim1 replied:
It's not working because `TypeError: igamma(): argument 'input' (position 1) must be Tensor, not float`.
I'm not sure how to make torch function to receive bare scalar (without converting it manually to Tensor first).
Or did you mean scalar tensor?

mruberry replied:
Oh right, sorry. Some of our functions accept scalars and I should have been clearer. For many of them we have to add explicit scalar overloads.

What about on CUDA, does putting a scalar in work there?

mfkasim1 replied:
No, it doesn't work either.

mruberry replied:
OK. Sounds good. Please ignore this suggestion, then.

mruberry commented on aten/src/ATen/native/cuda/Math.cuh:
```diff
@@ -174,6 +174,504 @@ static inline __host__ __device__ scalar_t calc_polygamma(int n, scalar_t x) {
 }
 
+/*
+ * The implementation of incomplete gamma function as well as the helper
```
"This implementation of the regularized incomplete gamma functions and their helper functions are derived from the implementation of SciPy's gammainc, Cephes's igam and igamc, and Boost's Lanczos approximations. See NOTICE for the licenses."

mruberry commented on aten/src/ATen/native/Math.h:
```diff
@@ -381,6 +381,717 @@ static inline float calc_polygamma(int64_t n, float x) {
+ */
+
+/*
+ * The implementation of incomplete gamma function as well as the helper
```
See updates to this comment from the .cuh file.

mruberry commented:
Hey @mfkasim1!

This is excellent work that's now only adding the regularized lower incomplete gamma function, but also adding the SciPy license and improving how our Cephes license and Boost license are cited. I made a last few small comments about test coverage and wording. The test failures on the PR seem unrelated to your work. Please ping me when this is ready to go! I'm looking forward to making it a part of PyTorch.

mfkasim1 committed:
Add more tests and update license NOTICE

mfkasim1 committed:
Making flake8 happy again

mfkasim1 commented:
Hi @mruberry. Thanks for all your comments. I believe your last comments have been addressed in the recent commits. The errors in CI are also unrelated to my work. So if you don't have any more comments, I think this is ready to go.

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6496,6 +6496,21 @@
+  use_c10_dispatcher: full
+  variants: method, function
+  dispatch:
+    DefaultBackend: igamma
```
I think this should be CPU, CUDA: here and below, following the updated note. Otherwise this will attempt to dispatch for XLA, which is probably why those odd XLA errors were appearing (an error would still appear, but it should be clearer if XLA doesn't have the function implemented).

mfkasim1 replied:
Done

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3286,6 +3286,41 @@ def merge_dicts(*dicts):
+.. math::
+    \text{out}_{i} = \frac{1}{\Gamma(\text{input}_i)} \int_0^{\text{other}_i} t^{\text{input}_i-1} e^{-t} dt
+
+where both :math:`\text{input}_i` and :math:`\text{other}_i` are weakly positive
```
This still needs to define the \\Gamma function used above. 

"where `\Gamma` is the ..." followed by its mathematical formula and then "and both ..." as is currently written. That is, the definition of \\Gamma should be inserted just after the mathematical expression.

mfkasim1 replied:
I put the definition of gamma later because I think it reads better.

mruberry replied:
Smart. It looks really good.

mruberry reviewed (commented):
Just two minor issues:

- (I think) dispatch needs an update based on the latest in the README
- doc needs to define the gamma function

mfkasim1 committed:
Update igamma dispatchers and docstring

mfkasim1 committed:
Update igamma docstring

mfkasim1 committed:
Merge remote-tracking branch 'upstream/master'

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3283,6 +3283,46 @@ def merge_dicts(*dicts):
+    \Gamma(\text{input}_i) = \int_0^\infty t^{\text{input}_i-1} e^{-t} dt.
+
+See :func:`torch.lgamma` for a related function.
+Supports :ref:`broadcasting to a common shape <broadcasting-semantics>`
```
nit: linebreak here so the "Supports..." statement is its own paragraph.

mfkasim1 replied:
This is also done

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3283,6 +3283,46 @@ def merge_dicts(*dicts):
+:math:`\Gamma(\cdot)` in the equation above is the gamma function,
+
+.. math::
+    \Gamma(\text{input}_i) = \int_0^\infty t^{\text{input}_i-1} e^{-t} dt.
```
nit:

\Gamma(\text{input}_i) = \int_0^\infty t^{(\text{input}_i-1)} e^{-t} dt.

This just adds parens around the exponential, which might be a little more readable.

mfkasim1 replied:
Done

mruberry commented:
Hey @mfkasim1, I know you didn't request another review but I took a quick look and made two very minor suggestions. Just let me know when you'd like this to be merged!

mfkasim1 committed:
Minor igamma docstring update

mruberry reviewed (approved):
Nice work, @mfkasim1! Thanks for taking the time to help with the licenses, too.

mruberry commented on aten/src/ATen/cpu/vec256/vec256_float_neon.h:
```diff
@@ -362,6 +362,16 @@ template <> class Vec256<float> {
+    store(tmp);
+    x.store(tmp_x);
+    for (int64_t i = 0; i < size(); i++) {
+      tmp[i] = calc_igamma(tmp[i], tmp_b[i]);
```
Internal build error:
```
xplat/caffe2/aten/src/ATen/cpu/vec256/vec256_float_neon.h:371:36: error: use of undeclared identifier 'tmp_b'; did you mean 'tmp_x'?
      tmp[i] = calc_igamma(tmp[i], tmp_b[i]);
```

I recommend removing neon support.

mfkasim1 replied:
I've removed the code in neon

mfkasim1 committed:
Removing support for neon

mfkasim1 commented:
@mruberry What happened in fb internal test? Is it not possible to expose the test?

Commit:
Attempted build fix

mruberry commented:
My build fix suggestion was incorrect. Sorry about that @mfkasim1. Trying another fix.

mfkasim1 committed:
Merge remote-tracking branch 'upstream/viable/strict'

mfkasim1 committed:
Removed tabs in igamma neon

mfkasim1 commented:
> My build fix suggestion was incorrect. Sorry about that @mfkasim1. Trying another fix.

Ah, no worries. I've added 2 commits to remove the tabs and merge `viable/upstream`

mfkasim1 commented:
@mruberry it still fails on facebook's internal test ...

mruberry commented:
> @mruberry it still fails on facebook's internal test ...

I know. Sorry. We're sorting out an internal test issue that's affecting multiple PRs and is not directly related to this PR. I'll update as soon as I get a signal from it.

mruberry commented:
Internal tests are passing. Just need to wait for this to be merged.

mfkasim1 commented:
Thanks @mruberry 👍

mfkasim1 commented:
Thanks @mruberry for your reviews and comments in this PR. Would it still be useful to do `torch.igammac` (the upper one) or should we just leave it to the users to do `igammac = 1 - igamma`?

mruberry commented:
That's a good question. A compromise approach would be to add a Note to the igamma documentation describing how to compute the regularized upper incomplete gamma function from it and see if we get a reguest for igammac itself. The example could also be extended with a comment, like, "Computes the regularized upper incomplete gamma function..."

Is the regularized upper incomplete gamma function used for anything? Bonus question, should we extend the igamma examples anyway with an example of it being used in context?

mfkasim1 commented:
The only advantage I see is that it might offer higher precision than `1 - igamma`. I'm not sure where upper igamma is used, anyway (my application only uses the lower one).
On the other hand, implementing it is not a big deal.
The function is there already (it's used to compute the lower igamma), so it just needs to dispatch the function to pytorch's API.
It's just the question of "can there be too many API functions?"

mruberry commented:
> The only advantage I see is that it might offer higher precision than `1 - igamma`. I'm not sure where upper igamma is used, anyway (my application only uses the lower one).

That's true, and it seems silly to have the real thing but not offer it if there are precision concerns. 

> On the other hand, implementing it is not a big deal.
> The function is there already (it's used to compute the lower igamma), so it just needs to dispatch the function to pytorch's API.

OK, let's just do it, then? 

> It's just the question of "can there be too many API functions?"

Yes, but this one is extremely low cost (for the reason you mention above). 


I appreciate your thoughtfulness, @mfkasim1. I suppose we might as well add it.

mfkasim1 commented:
Great, thanks for your answers! I'll do it soon.
